File: Patterns/Command Pattern/CSharp/Command/CailingFan/CailingFanOnToLowCommand.cs
namespace Command.CailingFan
{
    public class CailingFanOnToLowCommand : ICommand
    {
        private CailingFan cailingFan;
        private int preview;

        public CailingFanOnToLowCommand(CailingFan cailingFan)
        {
            preview = CailingFan.OFF;
            this.cailingFan = cailingFan;
        }

        public void Execute()
        {
            preview = cailingFan.State;
            cailingFan.SetLow();
        }

        public void Undo()
        {
            switch (preview)
            {
                case CailingFan.HIGH:
                    cailingFan.SetHigh();
                    break;
                case CailingFan.MEDIUM:
                    cailingFan.SetMedium();
                    break;
                case CailingFan.OFF:
                    cailingFan.Off();
                    break;
            }
        }
    }
}

File: Patterns/Command Pattern/Cpp/Command/CailingFan/CailingFanOnToLowCommand.cpp
#include "../ICommand.cpp"
#include "./CailingFan.cpp"

class CailingFanOnToLowCommand : public ICommand {
private:
    CailingFan* cailingFan;
    int preview;

public:
    CailingFanOnToLowCommand(CailingFan* cailingFan) : cailingFan{cailingFan}, preview{cailingFan->OFF} {}

    void execute() override {
        preview = cailingFan->getState();
        cailingFan->setLow();
    }

    void undo() override {
        if (preview == cailingFan->HIGH) cailingFan->setHigh();
        else if (preview == cailingFan->LOW) cailingFan->setLow();
        else if (preview == cailingFan->OFF) cailingFan->off();
    }

    string getCommandName() override {
        return "CailingFanOnToLowCommand";
    }
};